File: macos/Sources/JasnostEnrollmentSecurity/SignedEnrollmentVerifier.swift
import CryptoKit
import Foundation
import JasnostCaptureCore

public struct SignedDeviceBundlePayload: Codable, Equatable, Sendable {
    public let schemaVersion: Int
    public let kind: String
    public let bundleId: String
    public let generation: Int
    public let issuer: String
    public let audience: String
    public let issuedAt: String
    public let bundleExpiresAt: String
    public let deviceId: String
    public let companyId: String
    public let areaId: String
    public let projectId: String
    public let stackURL: String
    public let archiveIngestURL: String
    public let token: String
    public let tokenId: String
    public let expiresAt: String
    public let tokenBucketScope: JazzArchiveTokenBucketScope
    public let sinkBucketId: String?
    public let componentAccess: [String]
    public let streamSourceId: String?
    public let streamEndpoint: String?

    public var deviceBundle: DeviceBundle {
        DeviceBundle(
            kind: kind,
            deviceId: deviceId,
            stackURL: stackURL,
            projectId: projectId,
            companyId: companyId,
            areaId: areaId,
            archiveIngestURL: archiveIngestURL,
            streamSourceId: streamSourceId,
            streamEndpoint: streamEndpoint,
            token: token,
            tokenId: tokenId,
            expiresAt: expiresAt,
            tokenBucketScope: tokenBucketScope,
            sinkBucketId: sinkBucketId,
            componentAccess: componentAccess)
    }
}

public struct AuthorizedSignedDeviceBundle: Sendable {
    public let payload: SignedDeviceBundlePayload
    public let envelopeDigest: String
    public let acceptance: EnrollmentAcceptanceDecision

    public var bundle: DeviceBundle { payload.deviceBundle }
}

public enum SignedEnrollmentError: Error, Equatable, CustomStringConvertible {
    case trustUnavailable
    case malformedEnvelope
    case invalidBase64URL
    case nonCanonicalProtectedHeader
    case invalidProtectedHeader
    case unsupportedAlgorithm
    case unknownKey
    case invalidSignature
    case nonCanonicalPayload
    case invalidPayload
    case issuerMismatch
    case audienceMismatch
    case notYetValid
    case bundleExpired
    case credentialExpired
    case rollback
    case collision
    case acceptanceStateUnavailable

    public var description: String {
        switch self {
        case .trustUnavailable:
            "No trusted Jazz enrollment issuer is configured on this Mac."
        case .malformedEnvelope:
            "The enrollment bundle is not a flattened signed Jazz bundle."
        case .invalidBase64URL:
            "The signed enrollment bundle contains invalid base64url."
        case .nonCanonicalProtectedHeader:
            "The signed enrollment protected header is not canonical JSON."
        case .invalidProtectedHeader:
            "The signed enrollment protected header is not the required Jazz profile."
        case .unsupportedAlgorithm:
            "The signed enrollment bundle does not use EdDSA."
        case .unknownKey:
            "The enrollment signing key is not trusted by this Mac."
        case .invalidSignature:
            "The enrollment bundle signature is invalid."
        case .nonCanonicalPayload:
            "The signed enrollment payload is not canonical JSON."
        case .invalidPayload:
            "The signed enrollment payload does not satisfy the v2 contract."
        case .issuerMismatch:
            "The enrollment bundle was signed for a different Jazz issuer."
        case .audienceMismatch:
            "The enrollment bundle was signed for a different client audience."
        case .notYetValid:
            "The enrollment bundle was issued in the future."
        case .bundleExpired:
            "The copied enrollment bundle has expired; issue a new one."
        case .credentialExpired:
            "The scoped enrollment credential has expired; issue a new bundle."
        case .rollback:
            "This device has already accepted a newer enrollment generation."
        case .collision:
            "The enrollment generation or bundle id was reused with different signed content."
        case .acceptanceStateUnavailable:
            "The local enrollment replay state is unavailable; import was stopped safely."
        }
    }
}

public struct SignedEnrollmentVerifier: Sendable {
    public static let expectedAlgorithm = "EdDSA"
    public static let expectedType = "application/jazz-device-bundle+jws"

    private static let envelopeKeys: Set<String> = ["protected", "payload", "signature"]
    private static let protectedKeys: Set<String> = ["alg", "kid", "typ"]
    private static let payloadKeys: Set<String> = [
        "schemaVersion", "kind", "bundleId", "generation", "issuer", "audience", "issuedAt",
        "bundleExpiresAt", "deviceId", "companyId", "areaId", "projectId", "stackURL",
        "archiveIngestURL", "token", "tokenId", "expiresAt", "tokenBucketScope",
        "sinkBucketId", "componentAccess", "streamSourceId", "streamEndpoint",
    ]
    private static let bundleIDPattern = try! NSRegularExpression(
        pattern: "^jdb_[a-f0-9]{32}$")
    private static let scopeIDPattern = try! NSRegularExpression(
        pattern: "^[a-z0-9][a-z0-9-]{0,63}$")
    private static let projectIDPattern = try! NSRegularExpression(pattern: "^[0-9]+$")
    private static let componentPattern = try! NSRegularExpression(
        pattern: "^[A-Za-z0-9][A-Za-z0-9._-]{0,255}$")
    public let trustPolicy: EnrollmentTrustPolicy

    public init(trustPolicy: EnrollmentTrustPolicy) {
        self.trustPolicy = trustPolicy
    }

    public func verify(_ text: String, now: Date = Date()) throws
        -> AuthorizedSignedDeviceBundle
    {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard
            !trimmed.isEmpty,
            trimmed.utf8.count <= 200_000,
            let envelopeData = trimmed.data(using: .utf8),
            StrictJSON.hasUniqueObjectKeys(envelopeData),
            let envelope = try? JSONSerialization.jsonObject(with: envelopeData)
                as? [String: Any],
            Set(envelope.keys) == Self.envelopeKeys,
            let protectedSegment = envelope["protected"] as? String,
            let payloadSegment = envelope["payload"] as? String,
            let signatureSegment = envelope["signature"] as? String
        else {
            throw SignedEnrollmentError.malformedEnvelope
        }

        guard
            protectedSegment.utf8.count <= 2_048,
            payloadSegment.utf8.count <= 131_072,
            signatureSegment.utf8.count == 86,
            let protectedData = EnrollmentEncoding.decodeBase64URL(
                protectedSegment,
                maximumBytes: 1_536),
            let payloadData = EnrollmentEncoding.decodeBase64URL(
                payloadSegment,
                maximumBytes: 98_304),
            let signature = EnrollmentEncoding.decodeBase64URL(
                signatureSegment,
                maximumBytes: 64),
            signature.count == 64
        else {
            throw SignedEnrollmentError.invalidBase64URL
        }

        guard StrictJSON.hasUniqueObjectKeys(protectedData) else {
            throw SignedEnrollmentError.invalidProtectedHeader
        }
        guard
            let protected = try? JSONSerialization.jsonObject(with: protectedData)
                as? [String: Any],
            EnrollmentEncoding.canonicalJSONObject(protected) == protectedData
        else {
            throw SignedEnrollmentError.nonCanonicalProtectedHeader
        }
        guard Set(protected.keys) == Self.protectedKeys else {
            throw SignedEnrollmentError.invalidProtectedHeader
        }
        guard protected["alg"] as? String == Self.expectedAlgorithm else {
            throw SignedEnrollmentError.unsupportedAlgorithm
        }
        guard
            protected["typ"] as? String == Self.expectedType,
            let keyID = protected["kid"] as? String,
            EnrollmentEncoding.isValidKeyID(keyID)
        else {
            throw SignedEnrollmentError.invalidProtectedHeader
        }
        guard let publicKey = trustPolicy.publicKey(for: keyID) else {
            throw SignedEnrollmentError.unknownKey
        }

        let signingInput = Data("\(protectedSegment).\(payloadSegment)".utf8)
        guard publicKey.isValidSignature(signature, for: signingInput) else {
            throw SignedEnrollmentError.invalidSignature
        }

        guard StrictJSON.hasUniqueObjectKeys(payloadData) else {
            throw SignedEnrollmentError.invalidPayload
        }
        guard
            let payloadObject = try? JSONSerialization.jsonObject(with: payloadData)
                as? [String: Any],
            EnrollmentEncoding.canonicalJSONObject(payloadObject) == payloadData
        else {
            throw SignedEnrollmentError.nonCanonicalPayload
        }
        guard Set(payloadObject.keys) == Self.payloadKeys else {
            throw SignedEnrollmentError.invalidPayload
        }
        let payload: SignedDeviceBundlePayload
        do {
            payload = try JSONDecoder().decode(SignedDeviceBundlePayload.self, from: payloadData)
        } catch {
            throw SignedEnrollmentError.invalidPayload
        }
        try validate(payload, now: now)

        guard
            let canonicalEnvelope = EnrollmentEncoding.canonicalJSONObject([
                "protected": protectedSegment,
                "payload": payloadSegment,
                "signature": signatureSegment,
            ])
        else {
            throw SignedEnrollmentError.malformedEnvelope
        }
        let envelopeDigest = SHA256.hash(data: canonicalEnvelope)
            .map { String(format: "%02x", $0) }
            .joined()
        return AuthorizedSignedDeviceBundle(
            payload: payload,
            envelopeDigest: envelopeDigest,
            acceptance: .pending)
    }

    private func validate(_ payload: SignedDeviceBundlePayload, now: Date) throws {
        guard
            payload.schemaVersion == 2,
            payload.kind == DeviceBundle.expectedKind,
            Self.matches(payload.bundleId, pattern: Self.bundleIDPattern),
            payload.generation >= 1,
            payload.generation <= 9_007_199_254_740_991,
            payload.issuer.unicodeScalars.count <= 2_048,
            payload.audience.unicodeScalars.count <= 256,
            Self.matches(payload.deviceId, pattern: Self.scopeIDPattern),
            Self.matches(payload.companyId, pattern: Self.scopeIDPattern),
            Self.matches(payload.areaId, pattern: Self.scopeIDPattern),
            Self.matches(payload.projectId, pattern: Self.projectIDPattern),
            payload.stackURL.unicodeScalars.count <= 2_048,
            payload.archiveIngestURL.unicodeScalars.count <= 2_048,
            !payload.token.isEmpty,
            payload.token.unicodeScalars.count <= 8_192,
            !payload.tokenId.isEmpty,
            payload.tokenId.unicodeScalars.count <= 256,
            payload.componentAccess.count <= 128,
            payload.componentAccess == payload.componentAccess.sorted(),
            Set(payload.componentAccess).count == payload.componentAccess.count,
            payload.componentAccess.allSatisfy({
                Self.matches($0, pattern: Self.componentPattern)
            }),
            payload.streamSourceId.map({
                !$0.isEmpty && $0.unicodeScalars.count <= 512
            }) ?? true,
            (payload.streamEndpoint?.unicodeScalars.count ?? 0) <= 8_192
        else {
            throw SignedEnrollmentError.invalidPayload
        }

        guard payload.issuer == trustPolicy.issuer else {
            throw SignedEnrollmentError.issuerMismatch
        }
        guard payload.audience == trustPolicy.audience else {
            throw SignedEnrollmentError.audienceMismatch
        }
        guard
            EnrollmentURLPolicy.isSecureOrigin(payload.issuer),
            KeboolaStack.normalize(payload.stackURL) == payload.stackURL,
            JazzArchiveControlPlaneURL.normalize(payload.archiveIngestURL)
                == payload.archiveIngestURL
        else {
            throw SignedEnrollmentError.invalidPayload
        }
        if let streamEndpoint = payload.streamEndpoint {
            guard EnrollmentURLPolicy.isSecureEndpoint(streamEndpoint) else {
                throw SignedEnrollmentError.invalidPayload
            }
        }
        if payload.streamSourceId != nil, payload.streamEndpoint == nil {
            throw SignedEnrollmentError.invalidPayload
        }

        switch payload.tokenBucketScope {
        case .sink:
            guard
                let sinkBucketID = payload.sinkBucketId,
                !sinkBucketID.isEmpty,
                sinkBucketID.unicodeScalars.count <= 512
            else {
                throw SignedEnrollmentError.invalidPayload
            }
        case .none:
            guard payload.sinkBucketId == nil else {
                throw SignedEnrollmentError.invalidPayload
            }
        }

        guard
            let issuedAt = Timestamps.parse(payload.issuedAt),
            let bundleExpiresAt = Timestamps.parse(payload.bundleExpiresAt),
            let credentialExpiresAt = Timestamps.parse(payload.expiresAt),
            issuedAt.timeIntervalSinceReferenceDate.isFinite,
            bundleExpiresAt.timeIntervalSinceReferenceDate.isFinite,
            credentialExpiresAt.timeIntervalSinceReferenceDate.isFinite,
            bundleExpiresAt > issuedAt,
            credentialExpiresAt >= bundleExpiresAt
        else {
            throw SignedEnrollmentError.invalidPayload
        }
        if issuedAt > now.addingTimeInterval(trustPolicy.clockSkew) {
            throw SignedEnrollmentError.notYetValid
        }
        let earliestAcceptedNow = now.addingTimeInterval(-trustPolicy.clockSkew)
        if earliestAcceptedNow >= bundleExpiresAt {
            throw SignedEnrollmentError.bundleExpired
        }
        if earliestAcceptedNow >= credentialExpiresAt {
            throw SignedEnrollmentError.credentialExpired
        }
    }

    private static func matches(_ value: String, pattern: NSRegularExpression) -> Bool {
        let range = NSRange(value.startIndex..<value.endIndex, in: value)
        return pattern.firstMatch(in: value, range: range)?.range == range
    }
}

File: windows/Sources/JazzEnrollmentSecurity/SignedEnrollmentVerifier.cs
using System.Text;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;
using JazzCaptureCore;
using JazzCaptureCore.Enrollment;

namespace JazzEnrollmentSecurity;

/// <summary>The v2 signed enrollment payload, exactly as the issuer signed it.</summary>
public sealed record SignedDeviceBundlePayload(
    long SchemaVersion,
    string Kind,
    string BundleId,
    long Generation,
    string Issuer,
    string Audience,
    string IssuedAt,
    string BundleExpiresAt,
    string DeviceId,
    string CompanyId,
    string AreaId,
    string ProjectId,
    string StackUrl,
    string ArchiveIngestUrl,
    string Token,
    string TokenId,
    string ExpiresAt,
    JazzArchiveTokenBucketScope TokenBucketScope,
    string? SinkBucketId,
    IReadOnlyList<string> ComponentAccess,
    string? StreamSourceId,
    string? StreamEndpoint)
{
    /// <summary>The routing tuple and credential this payload projects into.</summary>
    public DeviceBundle DeviceBundle => new(
        Kind,
        DeviceId,
        StackUrl,
        ProjectId,
        CompanyId,
        AreaId,
        ArchiveIngestUrl,
        StreamSourceId,
        StreamEndpoint,
        Token,
        TokenId,
        ExpiresAt,
        TokenBucketScope,
        SinkBucketId,
        ComponentAccess);
}

/// <summary>A payload that passed signature, time, scope and replay admission.</summary>
public sealed record AuthorizedSignedDeviceBundle(
    SignedDeviceBundlePayload Payload,
    string EnvelopeDigest,
    EnrollmentAcceptanceDecision Acceptance)
{
    /// <summary>The routing tuple and credential carried by the authorized payload.</summary>
    public DeviceBundle Bundle => Payload.DeviceBundle;
}

/// <summary>
/// Verifies a flattened Ed25519 JWS enrollment bundle against a code-signed issuer policy.
/// </summary>
/// <remarks>
/// Nothing here trusts a value carried by the document itself. The algorithm is pinned, the key is
/// selected out of the code-signed anchor set by <c>kid</c> only, the issuer and audience must equal
/// the configured ones, and both the protected header and the payload must be byte-identical to
/// their own canonical re-serialization so there is exactly one spelling of any signed document.
/// </remarks>
public sealed partial class SignedEnrollmentVerifier
{
    /// <summary>The only signature algorithm accepted.</summary>
    public const string ExpectedAlgorithm = "EdDSA";

    /// <summary>The only protected <c>typ</c> accepted.</summary>
    public const string ExpectedType = "application/jazz-device-bundle+jws";

    private static readonly string[] EnvelopeKeys = { "protected", "payload", "signature" };
    private static readonly string[] ProtectedKeys = { "alg", "kid", "typ" };

    private static readonly string[] PayloadKeys =
    {
        "schemaVersion", "kind", "bundleId", "generation", "issuer", "audience", "issuedAt",
        "bundleExpiresAt", "deviceId", "companyId", "areaId", "projectId", "stackURL",
        "archiveIngestURL", "token", "tokenId", "expiresAt", "tokenBucketScope",
        "sinkBucketId", "componentAccess", "streamSourceId", "streamEndpoint",
    };

    /// <summary>Creates a verifier bound to <paramref name="trustPolicy"/>.</summary>
    public SignedEnrollmentVerifier(EnrollmentTrustPolicy trustPolicy) => TrustPolicy = trustPolicy;

    /// <summary>The code-signed trust this verifier answers to.</summary>
    public EnrollmentTrustPolicy TrustPolicy { get; }

    /// <summary>
    /// Verifies <paramref name="text"/> and returns the payload with acceptance still
    /// <see cref="EnrollmentAcceptanceDecision.Pending"/>. Durable replay admission is a separate,
    /// later step; this result must not reach anything token-bearing on its own.
    /// </summary>
    /// <exception cref="SignedEnrollmentException">The bundle was refused.</exception>
    public AuthorizedSignedDeviceBundle Verify(string text, DateTimeOffset now)
    {
        string trimmed = text.Trim();
        byte[] envelopeBytes;
        try
        {
            envelopeBytes = new UTF8Encoding(false, true).GetBytes(trimmed);
        }
        catch (EncoderFallbackException)
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.MalformedEnvelope);
        }

        if (envelopeBytes.Length == 0 || envelopeBytes.Length > 200_000)
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.MalformedEnvelope);
        }

        JsonObject? envelope = StrictJson.TryParseObject(envelopeBytes);
        if (envelope is null || !StrictJson.HasExactlyKeys(envelope, EnvelopeKeys))
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.MalformedEnvelope);
        }

        string? protectedSegment = StrictJson.StringOrNull(envelope, "protected");
        string? payloadSegment = StrictJson.StringOrNull(envelope, "payload");
        string? signatureSegment = StrictJson.StringOrNull(envelope, "signature");
        if (protectedSegment is null || payloadSegment is null || signatureSegment is null)
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.MalformedEnvelope);
        }

        byte[]? protectedBytes = protectedSegment.Length <= 2_048
            ? EnrollmentEncoding.DecodeBase64Url(protectedSegment, maximumBytes: 1_536)
            : null;
        byte[]? payloadBytes = payloadSegment.Length <= 131_072
            ? EnrollmentEncoding.DecodeBase64Url(payloadSegment, maximumBytes: 98_304)
            : null;
        byte[]? signature = signatureSegment.Length == 86
            ? EnrollmentEncoding.DecodeBase64Url(signatureSegment, maximumBytes: 64)
            : null;
        if (protectedBytes is null || payloadBytes is null || signature is null || signature.Length != 64)
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.InvalidBase64Url);
        }

        JsonObject? protectedHeader = StrictJson.TryParseObject(protectedBytes);
        if (protectedHeader is null)
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.InvalidProtectedHeader);
        }

        byte[]? canonicalProtected = EnrollmentEncoding.TryCanonicalJson(protectedHeader);
        if (canonicalProtected is null || !canonicalProtected.AsSpan().SequenceEqual(protectedBytes))
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.NonCanonicalProtectedHeader);
        }

        if (!StrictJson.HasExactlyKeys(protectedHeader, ProtectedKeys))
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.InvalidProtectedHeader);
        }

        if (StrictJson.StringOrNull(protectedHeader, "alg") != ExpectedAlgorithm)
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.UnsupportedAlgorithm);
        }

        string? keyId = StrictJson.StringOrNull(protectedHeader, "kid");
        if (StrictJson.StringOrNull(protectedHeader, "typ") != ExpectedType
            || keyId is null
            || !EnrollmentEncoding.IsValidKeyId(keyId))
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.InvalidProtectedHeader);
        }

        if (!TrustPolicy.HasPublicKey(keyId))
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.UnknownKey);
        }

        byte[] signingInput = Encoding.ASCII.GetBytes(protectedSegment + "." + payloadSegment);
        if (!TrustPolicy.VerifySignature(keyId, signingInput, signature))
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.InvalidSignature);
        }

        JsonObject? payloadObject = StrictJson.TryParseObject(payloadBytes);
        if (payloadObject is null)
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.InvalidPayload);
        }

        byte[]? canonicalPayload = EnrollmentEncoding.TryCanonicalJson(payloadObject);
        if (canonicalPayload is null || !canonicalPayload.AsSpan().SequenceEqual(payloadBytes))
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.NonCanonicalPayload);
        }

        if (!StrictJson.HasExactlyKeys(payloadObject, PayloadKeys))
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.InvalidPayload);
        }

        SignedDeviceBundlePayload payload = Decode(payloadObject);
        Validate(payload, now);

        var canonicalEnvelope = new JsonObject
        {
            ["protected"] = protectedSegment,
            ["payload"] = payloadSegment,
            ["signature"] = signatureSegment,
        };
        byte[]? envelopeDigestBytes = EnrollmentEncoding.TryCanonicalJson(canonicalEnvelope);
        if (envelopeDigestBytes is null)
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.MalformedEnvelope);
        }

        return new AuthorizedSignedDeviceBundle(
            payload,
            EnrollmentEncoding.HexSha256(envelopeDigestBytes),
            EnrollmentAcceptanceDecision.Pending);
    }

    private static SignedDeviceBundlePayload Decode(JsonObject payload)
    {
        long? schemaVersion = StrictJson.IntegerOrNull(payload, "schemaVersion");
        long? generation = StrictJson.IntegerOrNull(payload, "generation");
        string? kind = StrictJson.StringOrNull(payload, "kind");
        string? bundleId = StrictJson.StringOrNull(payload, "bundleId");
        string? issuer = StrictJson.StringOrNull(payload, "issuer");
        string? audience = StrictJson.StringOrNull(payload, "audience");
        string? issuedAt = StrictJson.StringOrNull(payload, "issuedAt");
        string? bundleExpiresAt = StrictJson.StringOrNull(payload, "bundleExpiresAt");
        string? deviceId = StrictJson.StringOrNull(payload, "deviceId");
        string? companyId = StrictJson.StringOrNull(payload, "companyId");
        string? areaId = StrictJson.StringOrNull(payload, "areaId");
        string? projectId = StrictJson.StringOrNull(payload, "projectId");
        string? stackUrl = StrictJson.StringOrNull(payload, "stackURL");
        string? archiveIngestUrl = StrictJson.StringOrNull(payload, "archiveIngestURL");
        string? token = StrictJson.StringOrNull(payload, "token");
        string? tokenId = StrictJson.StringOrNull(payload, "tokenId");
        string? expiresAt = StrictJson.StringOrNull(payload, "expiresAt");
        JazzArchiveTokenBucketScope? scope =
            JazzArchiveTokenBucketScopeNames.TryParse(StrictJson.StringOrNull(payload, "tokenBucketScope"));
        IReadOnlyList<string>? componentAccess = StrictJson.StringArrayOrNull(payload, "componentAccess");

        if (schemaVersion is null || generation is null || kind is null || bundleId is null
            || issuer is null || audience is null || issuedAt is null || bundleExpiresAt is null
            || deviceId is null || companyId is null || areaId is null || projectId is null
            || stackUrl is null || archiveIngestUrl is null || token is null || tokenId is null
            || expiresAt is null || scope is null || componentAccess is null)
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.InvalidPayload);
        }

        return new SignedDeviceBundlePayload(
            schemaVersion.Value,
            kind,
            bundleId,
            generation.Value,
            issuer,
            audience,
            issuedAt,
            bundleExpiresAt,
            deviceId,
            companyId,
            areaId,
            projectId,
            stackUrl,
            archiveIngestUrl,
            token,
            tokenId,
            expiresAt,
            scope.Value,
            OptionalString(payload, "sinkBucketId"),
            componentAccess,
            OptionalString(payload, "streamSourceId"),
            OptionalString(payload, "streamEndpoint"));
    }

    /// <summary>
    /// Reads an optional member that the v2 contract requires to be present and either a string or
    /// the JSON literal <c>null</c>.
    /// </summary>
    /// <remarks>
    /// The explicit <c>null</c> is the server's wire shape, not something this client emits: an
    /// absent key is a different document and would already have failed the exact-key-set check.
    /// </remarks>
    private static string? OptionalString(JsonObject payload, string key)
    {
        if (StrictJson.IsExplicitNull(payload, key))
        {
            return null;
        }

        return StrictJson.StringOrNull(payload, key)
            ?? throw new SignedEnrollmentException(SignedEnrollmentError.InvalidPayload);
    }

    private void Validate(SignedDeviceBundlePayload payload, DateTimeOffset now)
    {
        bool structurallyValid =
            payload.SchemaVersion == 2
            && payload.Kind == DeviceBundle.ExpectedKind
            && BundleIdPattern().IsMatch(payload.BundleId)
            && payload.Generation >= 1
            && payload.Generation <= 9_007_199_254_740_991
            && EnrollmentEncoding.ScalarCount(payload.Issuer) <= 2_048
            && EnrollmentEncoding.ScalarCount(payload.Audience) <= 256
            && ScopeIdPattern().IsMatch(payload.DeviceId)
            && ScopeIdPattern().IsMatch(payload.CompanyId)
            && ScopeIdPattern().IsMatch(payload.AreaId)
            && ProjectIdPattern().IsMatch(payload.ProjectId)
            && EnrollmentEncoding.ScalarCount(payload.StackUrl) <= 2_048
            && EnrollmentEncoding.ScalarCount(payload.ArchiveIngestUrl) <= 2_048
            && payload.Token.Length > 0
            && EnrollmentEncoding.ScalarCount(payload.Token) <= 8_192
            && payload.TokenId.Length > 0
            && EnrollmentEncoding.ScalarCount(payload.TokenId) <= 256
            && payload.ComponentAccess.Count <= 128
            && IsSortedAndUnique(payload.ComponentAccess)
            && payload.ComponentAccess.All(component => ComponentPattern().IsMatch(component))
            && (payload.StreamSourceId is null
                || (payload.StreamSourceId.Length > 0
                    && EnrollmentEncoding.ScalarCount(payload.StreamSourceId) <= 512))
            && EnrollmentEncoding.ScalarCount(payload.StreamEndpoint ?? string.Empty) <= 8_192;
        if (!structurallyValid)
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.InvalidPayload);
        }

        if (payload.Issuer != TrustPolicy.Issuer)
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.IssuerMismatch);
        }

        if (payload.Audience != TrustPolicy.Audience)
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.AudienceMismatch);
        }

        if (!EnrollmentUrlPolicy.IsSecureOrigin(payload.Issuer)
            || KeboolaStack.Normalize(payload.StackUrl) != payload.StackUrl
            || JazzArchiveControlPlaneUrl.Normalize(payload.ArchiveIngestUrl) != payload.ArchiveIngestUrl)
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.InvalidPayload);
        }

        if (payload.StreamEndpoint is string endpoint && !EnrollmentUrlPolicy.IsSecureEndpoint(endpoint))
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.InvalidPayload);
        }

        if (payload.StreamSourceId is not null && payload.StreamEndpoint is null)
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.InvalidPayload);
        }

        switch (payload.TokenBucketScope)
        {
            case JazzArchiveTokenBucketScope.Sink
                when payload.SinkBucketId is null
                    || payload.SinkBucketId.Length == 0
                    || EnrollmentEncoding.ScalarCount(payload.SinkBucketId) > 512:
                throw new SignedEnrollmentException(SignedEnrollmentError.InvalidPayload);
            case JazzArchiveTokenBucketScope.None when payload.SinkBucketId is not null:
                throw new SignedEnrollmentException(SignedEnrollmentError.InvalidPayload);
            default:
                break;
        }

        DateTimeOffset? issuedAt = Timestamps.TryParseRfc3339(payload.IssuedAt);
        DateTimeOffset? bundleExpiresAt = Timestamps.TryParseRfc3339(payload.BundleExpiresAt);
        DateTimeOffset? credentialExpiresAt = Timestamps.TryParseRfc3339(payload.ExpiresAt);
        if (issuedAt is null
            || bundleExpiresAt is null
            || credentialExpiresAt is null
            || bundleExpiresAt <= issuedAt
            || credentialExpiresAt < bundleExpiresAt)
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.InvalidPayload);
        }

        if (issuedAt > now + TrustPolicy.ClockSkew)
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.NotYetValid);
        }

        DateTimeOffset earliestAcceptedNow = now - TrustPolicy.ClockSkew;
        if (earliestAcceptedNow >= bundleExpiresAt)
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.BundleExpired);
        }

        if (earliestAcceptedNow >= credentialExpiresAt)
        {
            throw new SignedEnrollmentException(SignedEnrollmentError.CredentialExpired);
        }
    }

    private static bool IsSortedAndUnique(IReadOnlyList<string> values)
    {
        for (int i = 1; i < values.Count; i++)
        {
            if (string.CompareOrdinal(values[i - 1], values[i]) >= 0)
            {
                return false;
            }
        }

        return true;
    }

    [GeneratedRegex("^jdb_[a-f0-9]{32}$")]
    private static partial Regex BundleIdPattern();

    [GeneratedRegex("^[a-z0-9][a-z0-9-]{0,63}$")]
    private static partial Regex ScopeIdPattern();

    [GeneratedRegex("^[0-9]+$")]
    private static partial Regex ProjectIdPattern();

    [GeneratedRegex("^[A-Za-z0-9][A-Za-z0-9._-]{0,255}$")]
    private static partial Regex ComponentPattern();
}
